File: java/SelectionSort/src/SelectionSort.java
public class SelectionSort {
    private SelectionSort() {
    }
    // 稳定的O(n2)
    public static <E extends Comparable<E>> void sort(E[] arr) {
        // arr[0...i) 是有序的， arr[i...n)  是无序的
        for (int i = 0; i < arr.length; i++) {
            // 选择 arr[i...n)中的最小值对应的索引
            int minIndex = i;
            for (int j = i; j < arr.length; j++) {
                if (arr[j].compareTo(arr[minIndex]) < 0) {
                    minIndex = j;
                }
            }
            swap(arr, i, minIndex);

        }
    }

    private static <E> void swap(E[] arr, int i, int j) {
        E t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }

    public static void main(String[] args) {
        int[] dataSize = {10000, 100000};
        for (int n : dataSize) {
            Integer[] arr = ArrayGenerator.generateRandomArray(n, n);
            SortingHelper.sortTest("SelectionSort", arr);
        }
    }
}

File: tsVersion/SelectionSort/selectionSort.ts
import {generateRandomArray} from "../SortingHelper/arrayGenerator";
import sortTest from "../SortingHelper/sortingHelper";

export const selectionSort = (arr: number[]) => {
  // arr[0..i]是有序的，arr[i..n]是无序的
  for (let i = 0; i < arr.length; i++) {
    let minIndex = i;
    for (let j = i; j < arr.length; j++) {
      if (arr[minIndex] > arr[j]) {
        swap(arr, minIndex, j)
      }
    }
  }
}
const swap = (arr: number[], minIndex: number, j: number) => {
  let t = arr[minIndex];
  arr[minIndex] = arr[j];
  arr[j] = t;
}


const dataSize = [10000, 100000]
for (const n of dataSize ) {
  const arr = generateRandomArray(n, n);
  sortTest('selectionSort', arr)
}